### PR TITLE
Lock when working with file references [run-systemtest]

### PR DESCRIPTION
**Pull request by hmusum**

### CODE REVIEW

File: configserver/src/main/java/com/yahoo/vespa/config/server/filedistribution/FileDBRegistry.java [hunk 1]
```diff
@@ -33,16 +33,21 @@ public class FileDBRegistry implements FileRegistry {
...
+        fileReferenceCache.putAll(knownReferences);
     }
 
     public static FileDBRegistry create(AddFileInterface manager, Reader persistedState) {
         try (BufferedReader reader = new BufferedReader(persistedState)) {
             String ignoredFileSourceHost = reader.readLine();
             if (ignoredFileSourceHost == null)
                 throw new RuntimeException("No file source host");
-            return new FileDBRegistry(manager, decode(reader));
+            return new FileDBRegistry(manager, decode(reader), false);
```
Comment (jonmv) [inline configserver/src/main/java/com/yahoo/vespa/config/server/filedistribution/FileDBRegistry.java:50]: This used to be `true`?

Comment (hmusum) [reply configserver/src/main/java/com/yahoo/vespa/config/server/filedistribution/FileDBRegistry.java:50]: Ouch, you are right, will fix

File: configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java [hunk 2]
```diff
@@ -594,22 +595,16 @@ public HttpResponse fileDistributionStatus(ApplicationId applicationId, Duration
...
-        List<String> fileReferencesToDelete = sortedUnusedFileReferences(fileReferencesPath, fileReferencesInUse, instant);
+        List<String> fileReferencesToDelete = sortedUnusedFileReferences(fileDirectory.getRoot(), fileReferencesInUse, instant);
         if (fileReferencesToDelete.size() > 0) {
             log.log(Level.FINE, () -> "Will delete file references not in use: " + fileReferencesToDelete);
-            fileReferencesToDelete.forEach(fileReference -> {
-                File file = new File(fileReferencesPath, fileReference);
-                if ( ! IOUtils.recursiveDeleteDir(file))
-                    log.log(Level.WARNING, "Could not delete " + file.getAbsolutePath());
-            });
+            fileReferencesToDelete.forEach(fileReference -> fileDirectory.delete(new FileReference(fileReference)));
```
Comment (jonmv) [inline configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java:607]: Won't you need to hold the lock while evaluating whether or not to delete the file, too?
The simplest is perhaps to pass a condition to evaluate inside the delete method, and re-run it while holding the lock there?

File: configserver/src/test/java/com/yahoo/vespa/config/server/filedistribution/FileDBRegistryTestCase.java [hunk 3]
```diff
@@ -79,7 +78,11 @@ public void importAndExport() throws IOException {
         checkConsistentEntry(fileRegistry.export().get(0), importedRegistry);
         checkConsistentEntry(fileRegistry.export().get(1), importedRegistry);
 
-        assertEquals(new FileReference("non-existing-file"), importedRegistry.addFile(NO_FOO_FILE));
+        try {
+            importedRegistry.addFile(NO_FOO_FILE);
+        } catch (Exception e ) {
+            assertEquals("java.io.FileNotFoundException: src/test/apps/zkapp/files/no_foo.json (No such file or directory)", e.getMessage());
```
[the remark this answers is not in the document]
Comment (hmusum) [reply configserver/src/test/java/com/yahoo/vespa/config/server/filedistribution/FileDBRegistryTestCase.java:86]: No, I thought this was caused by another change, but this should be reverted

File: configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java (hunk 2 above)
Comment (hmusum) [reply configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java]: Yeah, that's right, working on it.

**Comment by hmusum** [comment]
@jonmv please take a look again

**Review by jonmv — APPROVED** [review]

**Pull request merged**